pkg/app/ops/firestoreindexcreator/gcloud.go

```
@@ -0,0 +1,104 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package firestoreindexcreator
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+
+	"go.uber.org/zap"
+)
+
+const defaultGcloudPath = "gcloud"
+
+type gcloud struct {
+	// The path to executable of gcloud.
+	gcloudPath string
+	// The Google Cloud Platform project ID
+	projectID string
+	// The path to the service account key file.
+	serviceAccountFile string
+
+	logger *zap.Logger
+}
+
+func newGcloud(gcloudPath, projectID, serviceAcccountFile string, logger *zap.Logger) *gcloud {
+	if gcloudPath == "" {
+		gcloudPath = defaultGcloudPath
+	}
+	return &gcloud{
+		gcloudPath:         gcloudPath,
+		projectID:          projectID,
+		serviceAccountFile: serviceAcccountFile,
+		logger:             logger.Named("gcloud-client"),
+	}
+}
+
+func (c *gcloud) authorize(ctx context.Context) error {
+	// Extracts client email from the service account file at first.
+	serviceAccount := struct {
+		ClientEmail string `json:"client_email"`
+	}{}
+	b, err := ioutil.ReadFile(c.serviceAccountFile)
+	if err != nil {
+		return fmt.Errorf("failed to open service account file: %w", err)
+	}
+	if err := json.Unmarshal(b, &serviceAccount); err != nil {
+		return fmt.Errorf("failed to parse service account: %w", err)
+	}
+	if serviceAccount.ClientEmail == "" {
+		return fmt.Errorf("unexpected service account file format")
+	}
+
+	return c.runGcloudCommand(ctx, "auth", "activate-service-account", serviceAccount.ClientEmail, "--key-file", c.serviceAccountFile)
+}
+
+func (c *gcloud) createIndex(ctx context.Context, idx *index) {
```

This function should return an error for handling by its caller.